In [1]:
from concurrent.futures import ProcessPoolExecutor
import math
from pathlib import Path
import re
from statistics import mean
import time
from typing import Iterator

from blist import sortedlist, blist
import numpy as np

from config import get_sessionmaker
from models import Page, PageLink, PageTalk, PageQuality

In [3]:
DATABASE_URI = "postgresql://postgres:postgres@localhost:5432/complete_wikipedia"
Session = get_sessionmaker(db_uri=DATABASE_URI)
s = Session()

In [3]:
results = s.query(
         Page, PageTalk, PageQuality,
    ).filter(
         PageQuality.page_id == PageTalk.page_id,
    ).filter(
         PageTalk.page_title == Page.page_title,
    ).all()

(<Page: (
	page_id=21345243,
	page_title='Marquis_de_Sade',
	)>, <PageTalk: (
	page_id=21345106,
	page_title='Marquis_de_Sade',
	)>, <PageQuality: (
	page_id=21345106,
	page_quality='B',
	)>)

In [15]:
for result in results:
    print(result)
    break

(<Page: (
	page_id=47419,
	page_title='Convenience_store',
	)>, <PageTalk: (
	page_id=47455,
	page_title='Convenience_store',
	)>, <PageQuality: (
	page_id=47455,
	page_quality='C',
	)>)


In [20]:
result_ids = []
for result in results:
    result_ids.append(result[0].page_id)
result_ids = np.array(result_ids)
np.save("data/high_quality_ids", result_ids)

In [4]:
result_ids = np.load("data/high_quality_ids.npy")
sl = sortedlist(result_ids)
del result_ids

In [5]:
%%timeit
47419 in sl

8.09 µs ± 78.8 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


In [6]:
%%timeit
47419 in result_ids

138 µs ± 102 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)


In [30]:
s.query(
         Page, PageTalk, PageQuality,
    ).filter(
         PageQuality.page_id == PageTalk.page_id,
    ).filter(
         PageTalk.page_title == Page.page_title,
    ).filter(
        Page.page_id == 25202,
    ).first()

(<Page: (
	page_id=25202,
	page_title='Quantum_mechanics',
	)>, <PageTalk: (
	page_id=19594240,
	page_title='Quantum_mechanics',
	)>, <PageQuality: (
	page_id=19594240,
	page_quality='GA',
	)>)

In [24]:
s.query(
         Page, PageTalk, PageQuality,
    ).filter(
         PageQuality.page_id == PageTalk.page_id,
    ).filter(
         PageTalk.page_title == Page.page_title,
    ).filter(
        Page.title_insensitive=="QuAnTum_mEchAnIcs",
    ).first()

(<Page: (
	page_id=25202,
	page_title='Quantum_mechanics',
	)>, <PageTalk: (
	page_id=19594240,
	page_title='Quantum_mechanics',
	)>, <PageQuality: (
	page_id=19594240,
	page_quality='GA',
	)>)

- 53008 lines
- 1,420,256,990 links (all ns)
- 626,327,844 links (ns 0 to all)
- 561,899,045 links (ns 0 to ns 0)
- at 6,301,566 ns 0 pages, 99.4 links/page (to all pages) & 89.1 links/page (to other ns 0 pages)


In [5]:
line_start = re.compile(b"INSERT INTO (.*?) VALUES ")
item_ex = re.compile(b"\((.*?),(.*?),'(.*?)',(.*?)\)")

In [6]:
def count_lines(line):
    return 1

def file_generator(file):
    for line in file:
        if line_start.search(line) is not None:
            yield line

def get_line_iterator(line) -> Iterator:
    if line_start.search(line) is not None:
        item_iter = item_ex.finditer(line)
        return item_iter
    else:
        return None


def get_items_in_line(line_iter):
    # sess_maker = get_sessionmaker(db_uri=DATABASE_URI)
    # sess = sess_maker()
    item_count = 0
    skip_count = 0
    links = []
    for entry in line_iter:
        if entry.group(4) == b"0" and entry.group(2) == b"0":
            pl_from = int(entry.group(1))
            pl_title = entry.group(3)
            if pl_from not in sl:
                skip_count += 1
                continue
            links.append([pl_from, pl_title])
            item_count += 1
    return [item_count, skip_count, links]


def process_file(path):
    with open(path, "rb") as f:
        line_iter = get_line_iterator(f.read())
        if line_iter is None:
            return None
        items = get_items_in_line(line_iter)
        return items
    

In [7]:
%%time
chunk_directory = Path("data/link_chunks")
item_counts = blist()
skip_counts = blist()
items = blist()
lookup = blist()
chunk_size = 7500
with ProcessPoolExecutor() as executor:
    for i, result in enumerate(executor.map(process_file, chunk_directory.iterdir())):
        if result is None:
            print("Error in regex for lines")
        item_counts.append(result[0])
        skip_counts.append(result[1])
        for item in result[2]:
            items.append(item)
        del result
        if i % chunk_size == 0:
            print(f"Counted {i} items:")
            avg_items = mean(item_counts[-chunk_size:])
            avg_skips = mean(skip_counts[-chunk_size:])
            print(f"chunk avg {avg_items} items")
            print(f"chunk avg {avg_skips} skips")
            start_time = time.time()
            print("starting sort")
            items.sort()
            delta = time.time() - start_time
            print(f"Finished sort in {delta/60} min")
        

Counted 0 items:
chunk avg 0 items
chunk avg 0 skips
starting sort
Finished sort in 2.6226043701171877e-07 min
Counted 7500 items:
chunk avg 1791.9105333333334 items
chunk avg 9390.250933333333 skips
starting sort
Finished sort in 0.630859661102295 min
Counted 15000 items:
chunk avg 1852.7834666666668 items
chunk avg 9542.773866666666 skips
starting sort
Finished sort in 1.0399975458780923 min
Counted 22500 items:
chunk avg 1935.7597333333333 items
chunk avg 9777.677466666666 skips
starting sort
Finished sort in 1.4385419209798178 min
Counted 30000 items:
chunk avg 1857.3285333333333 items
chunk avg 9567.483066666666 skips
starting sort
Finished sort in 1.622099268436432 min
Counted 37500 items:
chunk avg 1867.9193333333333 items
chunk avg 9466.9324 skips
starting sort
Finished sort in 1.9107801040013632 min
Counted 45000 items:
chunk avg 1850.3845333333334 items
chunk avg 9465.7932 skips
starting sort
Finished sort in 2.2516356309254966 min
Counted 52500 items:
chunk avg 1824.86666666

In [8]:
start_time = time.time()
print("starting sort")
items.sort()
delta = time.time() - start_time
print(f"Finished sort in {delta/60} min")

starting sort
Finished sort in 1.7812417030334473 min


In [9]:
%%time
start_time = time.time()
print("starting load")
item = items.pop()
current_idx = item[0]
current_page = [current_idx, [item[1].decode("utf-8").replace("_", " ")]]
pl_added = 0
while True:
    try:
        item = items.pop()
        if item[0] == current_idx:
            current_page[1].append(item[1].decode("utf-8").replace("_", " "))
        else:
            link = PageLink(
                pl_from=int(current_page[0]),
                pl_titles=current_page[1]
            )
            s.add(link)
            pl_added += 1
            if pl_added % 100 == 0:
                s.commit()
            if pl_added % 50000 == 0:
                print(f"added {pl_added} pages' links")
            current_idx = item[0]
            current_page = [current_idx, [item[1].decode("utf-8").replace("_", " ")]]
    except IndexError:
        link = PageLink(
            pl_from=current_page[0],
            pl_titles=current_page[1]
        )
        s.add(link)
        s.commit()
        break
delta = time.time() - start_time
print(f"Finished load in {delta/60} min")

starting load
added 50000 pages' links
added 100000 pages' links
added 150000 pages' links
added 200000 pages' links
added 250000 pages' links
added 300000 pages' links
added 350000 pages' links
added 400000 pages' links
added 450000 pages' links
added 500000 pages' links
Finished load in 25.91617283821106 min
CPU times: user 4min 24s, sys: 2.04 s, total: 4min 26s
Wall time: 25min 54s


In [ ]:
# chunk_directory = Path("data/chunks")
# item_counts = blist()
# skip_counts = blist()
# items = blist()
# lookup = blist()
# chunk_size = 2000
# with ProcessPoolExecutor() as executor:
#     for i, result in enumerate(executor.map(process_file, chunk_directory.iterdir())):
#         if result is None:
#             print("Error in regex for lines")
#         item_counts.append(result[0])
#         skip_counts.append(result[1])
#         for item in result[2]:
#             try:
#                 idx = lookup.index(item[1])
#                 items.append([item[0], idx])
#             except ValueError:
#                 lookup.append(item[1])
#                 idx = lookup.index(item[1])
#                 items.append([item[0], idx])
#         if i % chunk_size == 0:
#             print(f"Counted {i} items:")
#             avg_items = mean(item_counts[-chunk_size:])
#             avg_skips = mean(skip_counts[-chunk_size:])
#             print(f"chunk avg {avg_items} items")
#             print(f"chunk avg {avg_skips} skips")
#         del result

In [9]:
len(items)

1574945

To run through whole link set:
CPU times: user 17 s, sys: 1.73 s, total: 18.7 s
Wall time: 14min 22s

Running through and getting links:
CPU times: user 1min 36s, sys: 8.03 s, total: 1min 44s
Wall time: 15min 47s

In [ ]:
whilte True:
    try:
        row = 

In [33]:
items = sortedlist()

In [41]:
items.add(blist([4, 8]))

In [42]:
items.index()

sortedlist([blist([4, 1]), blist([4, 4]), blist([4, 5]), blist([4, 8])])

In [31]:
type(items[0])

list

In [35]:
561899045/6301566 

89.16815994627368

In [34]:
with open("data/enwiki-20210520-pagelinks.sql", "rb") as file:
    for i, line in enumerate(file_generator(file)):
        f = open(f"data/chunks/{i}.sql", "wb")
        f.write(line)
        f.close()

KeyboardInterrupt: 

PosixPath('data/chunks/28951.sql')

In [5]:
line_count

53008

In [ ]:


def test_func(a):
    return 1
test_list = [1,2,3,4,5,6,7]
with concurrent.futures.ProcessPoolExecutor() as executor:
    for result in executor.map(test_func, test_list)


In [45]:
%%time
num_qualities = s.query(PageQuality).count()
offset = 0
limit = 50000
offset_multipliers = range(math.ceil(num_qualities/limit))
retrieved = 0

def scan_db(offset_multiplier):
    with Session() as sess:
        offset = offset_multiplier * limit
        results = sess.query(PageQuality).offset(offset).limit(limit)
    return offset_multiplier #results.count()

# for i in offset_multipliers:
#     retrieved += scan_db(i, limit)
#     print(f"retrieved {retrieved} entries")
# args = ((multiplier, limit) for multiplier in offset_multipliers)
with concurrent.futures.ProcessPoolExecutor() as executor:
    for result_count in executor.map(scan_db, offset_multipliers):
        retrieved += result_count
        print(f"retrieved {retrieved} entries")

retrieved 0 entries
retrieved 1 entries
retrieved 3 entries
retrieved 6 entries
retrieved 10 entries
retrieved 15 entries
retrieved 21 entries
retrieved 28 entries
retrieved 36 entries
retrieved 45 entries
retrieved 55 entries
CPU times: user 13.4 ms, sys: 72.5 ms, total: 85.9 ms
Wall time: 116 ms


In [48]:
%%time
num_qualities = s.query(PageQuality).count()
offset = 0
limit = 50000
offset_multipliers = range(math.ceil(num_qualities/limit))
retrieved = 0

def scan_db(offset_multiplier):
    DATABASE_URI = "postgresql://postgres:postgres@localhost:5432/complete_wikipedia"
    sess_maker = get_sessionmaker(db_uri=DATABASE_URI)
    with sess_maker() as sess:
        offset = offset_multiplier * limit
        results = sess.query(PageQuality).offset(offset).limit(limit)
        for result in results:
            result.page_talk.page.page_title
    return results.count()

with concurrent.futures.ProcessPoolExecutor() as executor:
    for result_count in executor.map(scan_db, offset_multipliers):
        retrieved += result_count
        print(f"retrieved {retrieved} entries")

retrieved 50000 entries
retrieved 100000 entries
retrieved 150000 entries
retrieved 200000 entries
retrieved 250000 entries
retrieved 300000 entries
retrieved 350000 entries
retrieved 400000 entries
retrieved 450000 entries
retrieved 500000 entries
retrieved 522137 entries
CPU times: user 19.1 ms, sys: 71.9 ms, total: 91 ms
Wall time: 1min 7s


## Loop through db:
Series, 1 CPU:
- CPU times: user 4min 52s
- sys: 16.3 s
- total: 5min 8s
- Wall time: 6min 57s

Parallel, 12 CPUs:
- CPU times: user 19.1 ms
- sys: 71.9 ms
- total: 91 ms
- Wall time: 1min 7s

In [4]:
result.page_talk.page

<Page: (
	page_id=733981,
	page_title='Flecainide',
	)>